lib/Dialect/Handshake/HandshakeOps.cpp

```diff
@@ -29,16 +29,93 @@ using namespace circt::handshake;
 #include "mlir/IR/IntegerSet.h"
 #include "mlir/IR/PatternMatch.h"
 #include "mlir/Transforms/InliningUtils.h"
+#include "llvm/ADT/Any.h"
 #include "llvm/ADT/SetVector.h"
 #include "llvm/ADT/SmallBitVector.h"
 #include "llvm/Support/Debug.h"
 
+#define INDEX_WIDTH 32
+
 namespace circt {
 namespace handshake {
 #include "circt/Dialect/Handshake/HandshakeOps.inc"
 }
 } // namespace circt
 
+// Convert ValueRange to vectors
+std::vector<mlir::Value> toVector(mlir::ValueRange range) {
+  return std::vector<mlir::Value>(range.begin(), range.end());
+}
+
+// Returns whether the precondition holds for a general op to execute
+bool isReadyToExecute(std::vector<mlir::Value> &ins,
+                      std::vector<mlir::Value> &outs,
+                      llvm::DenseMap<mlir::Value, llvm::Any> &valueMap) {
+  for (auto in : ins)
+    if (valueMap.count(in) == 0)
+      return false;
+
+  for (auto out : outs)
+    if (valueMap.count(out) > 0)
+      return false;
+
+  return true;
+}
+
+// Fetch values from the value map and consume them
+std::vector<llvm::Any>
+fetchValues(std::vector<mlir::Value> &values,
```

Comment:
nit: could use `ArrayRef<mlir::Value>` for `std::vector<mlir::Value>&` (same for other func decl. in this file)